File: week1_baseline/python/07_the_run_dsl/boukensha/prompt_builder.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .backends.base import Base
    from .context import Context


class PromptBuilder:
    def __init__(self, context: Context, backend: Base) -> None:
        self.context = context
        self.backend = backend

    def to_messages(self) -> list[dict[str, Any]]:
        return self.backend.to_messages(self.context.messages)

    def to_tools(self) -> list[dict[str, Any]]:
        return self.backend.to_tools(self.context.tools)

    def to_api_payload(
        self, *, max_output_tokens: int = 1024, tools: list[dict[str, Any]] | None = None
    ) -> dict[str, Any]:
        return self.backend.to_payload(self.context, max_output_tokens=max_output_tokens, tools=tools)

    def parse_response(self, response: dict[str, Any]) -> dict[str, Any]:
        return self.backend.parse_response(response)

    @property
    def headers(self) -> dict[str, str]:
        return self.backend.headers

    @property
    def url(self) -> str:
        return self.backend.url

File: week1_baseline/ruby/12_context/lib/boukensha/prompt_builder.rb
module Boukensha
  class PromptBuilder
    attr_reader :backend

    def initialize(context, backend)
      @context = context
      @backend = backend
    end

    def to_messages
      @backend.to_messages(@context.messages)
    end

    def to_tools
      @backend.to_tools(@context.tools)
    end

    def to_api_payload(max_output_tokens: 1024, tools: nil)
      @backend.to_payload(@context, max_output_tokens: max_output_tokens, tools: tools)
    end

    # Delegates to the backend, which normalizes a provider response into the
    # common shape documented in Backends::Base:
    #   { stop_reason: "tool_use" | "end_turn",
    #     content: [ {"type"=>"reasoning",...} | {"type"=>"text",...} | {"type"=>"tool_use",...} ] }
    def parse_response(response)
      @backend.parse_response(response)
    end

    def headers
      @backend.headers
    end

    def url
      @backend.url
    end
  end
end